neqo-transport/src/shuffle.rs

```diff
@@ -0,0 +1,189 @@
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use std::ops::Range;
+
+use neqo_common::qtrace;
+
+/// Finds the range where the SNI extension lives.
+///
+/// If this isn't a `ClientHello` or the range cannot be found, return the whole message.
+fn find_sni(buf: &[u8]) -> Range<usize> {
+    // Read a big-endian integer from `buf`.
+    fn read_len(buf: &[u8]) -> usize {
+        let mut len = 0;
+        for v in buf {
+            len = (len << 8) + usize::from(*v);
+        }
+        len
+    }
+
+    // Advance `i` by the value read from the `N` bytes at `i`, and return the new index.
+    // If the buffer is too short, return `None`.
+    fn skip_vec<const N: usize>(i: usize, buf: &[u8]) -> Option<usize> {
+        if i + N > buf.len() {
+            return None;
+        }
+        let i = i + N + read_len(&buf[i..i + N]);
+        if i > buf.len() {
+            None
+        } else {
+            Some(i)
+        }
+    }
+
+    let mut i = 1 + 3 + 2 + 32; // msg_type, length, version, random
+
+    // Return if buf is too short or does not contain a ClientHello (first byte== 1)
+    if buf.len() < i || buf[0] != 1 {
+        return 0..buf.len();
+    }
+
+    // Skip session_id
+    i = if let Some(i) = skip_vec::<1>(i, buf) {
+        i
+    } else {
+        return 0..buf.len();
+    };
```

Comment:
I wonder if we can't instead do something like this:

```rust
i += skip_vec::<1>(i, buf)?;
```
With the function returning `Option<Range<usize>>` to allow for that.  The caller would need to generate the default, but that seems like a good trade.

Reply:
I was thinking to check if we can use/extend `neqo-common/src/codec.rs` here.